### PR TITLE
Avoid concurrent writes to a map in dynamicWatch

### LINKED ISSUE #193: panic: concurrent map writes in watch/dynamic.go
We're seeing a `panic` on [`dw.lastRV[key] = rv`](https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern/blob/7dcbbb6b50c8044fafa297997930d151152cde2a/pkg/patterns/declarative/pkg/watch/dynamic.go#L150) with the following stack trace:
```
"concurrent map writes"
Stack:
	2  0x000000000309d871 in sigs.k8s.io/kubebuilder-declarative-pattern/pkg/patterns/declarative/pkg/watch.(*dynamicWatch).watchUntilClosed
	    at /home/tomasl/go/pkg/mod/ghe.spotify.net/declarative-infra/kubebuilder-declarative-pattern@v0.0.4/pkg/patterns/declarative/pkg/watch/dynamic.go:150
	3  0x000000000309c71a in sigs.k8s.io/kubebuilder-declarative-pattern/pkg/patterns/declarative/pkg/watch.(*dynamicWatch).Add.func1
	    at /home/tomasl/go/pkg/mod/ghe.spotify.net/declarative-infra/kubebuilder-declarative-pattern@v0.0.4/pkg/patterns/declarative/pkg/watch/dynamic.go:86
```

It seems to happen every time our operator starts, but for different keys every time. I don't grok what goes on under the hood well enough to understand what different paths could be updating this simultaneously; what am I looking for in troubleshooting this? Or is it just a case of having too many resources? (This operator is looking at 1000+ instances of the CR, which each creates a couple of `RoleBinding`s and an [`IAMPartialPolicy`](https://cloud.google.com/config-connector/docs/reference/resource-docs/iam/iampartialpolicy) in the namespace where the CR is, plus a `RoleBinding` in a common namespace.)

If there's anything I can help out with to contribute a workaround/fix here, please let me know.

### PR DESCRIPTION
**Pull request by tomasaschan**
Fixes #193

CC @justinsb

### CODE REVIEW

**Comment by justinsb** [comment]
Thanks @tomasaschan 

/ok-to-test
/approve
/lgtm

**Comment by tomasaschan** [comment]
Ah, I'm not allowed to close issues with the comment? Let me rephrase then...

**Comment by tomasaschan** [comment]
@justinsb I think this needs another /lgtm, since I had to remove something from the commit message.
